### PR TITLE
Enable GT_CALL with long ret types for x86 RyuJIT

### PR DESCRIPTION
**Pull request by adiaaida**
1) Enables genMultiRegCallStoreToLocal for x86 RyuJIT. Forces long return
types to be stored to the stack after returning from a call.

2) Update lvaPromoteLongVars to not promote a long if it is a multi reg
arg or ret.

3) Adds NYI for call arguments that are longs and contain adds or
subtracts. We are currently producing the wrong order of operations so
that we can push the argument immediately after performing the arithmetic.
We will do the hi adc/sbb before the carry bit has been set by doing the
lo operation.

4) Adds an NYI for morphing a node into a call node if the call will have
a long return type.

5) Moves the logic for forcing var = call() for calls with long return
types to lower::DecomposeNode().

### CODE REVIEW

**Comment by adiaaida** [comment]
@CarolEidt @sivarv PTAL
/cc: @dotnet/jit-contrib

File: src/jit/lower.cpp [hunk 1]
```diff
@@ -381,8 +383,44 @@ void Lowering::DecomposeNode(GenTreePtr* pTree, Compiler::fgWalkData* data)
...
+                        comp->lvaTable[varNum].lvIsMultiRegArgOrRet = true;
+                        break;
+                    }
+                }
+
+                // Otherwise, we need to force var = call()
+                GenTreePtr* treePtr = nullptr;
+                parent = tree->gtGetParent(&treePtr);
+
+                GenTreeStmt* asgStmt = comp->fgInsertEmbeddedFormTemp(treePtr);
```
Comment (CarolEidt) [inline src/jit/lower.cpp]: I would suggest adding a TODO here to handle the possibility that the new asgStmt will become a top-level statement, carrying existing embedded statements with it. This would be a good candidate for using the `CreateTemporary` method in @mikedn's PR #5871

File: src/jit/morph.cpp [hunk 2]
```diff
@@ -83,6 +83,18 @@ GenTreePtr          Compiler::fgMorphIntoHelperCall(GenTreePtr      tree,
...
+    {
+        GenTreeCall* callNode = tree->AsCall();
+        ReturnTypeDesc* retTypeDesc = callNode->GetReturnTypeDesc();
+        retTypeDesc->Reset();
+        retTypeDesc->Initialize(this, callNode->gtRetClsHnd);
+        
+        NYI("Helper with TYP_LONG return type");
+    }
+#endif
+
```
Comment (CarolEidt) [inline src/jit/morph.cpp]: I'm curious why this is needed; is there something different about the helper call case, as compared with the generic call case that you are now handling? Is this because they will not have been forced to a var=call by the code that handles multi-reg structs? If so, then I'm wondering why the lowering code needs to handle the case where it is not already var=call. I suspect there's something I'm missing ...

Comment (adiaaida) [inline src/jit/morph.cpp]: There is an issue with the parameters to helper calls where sometimes (in about 5 tests), we try to decompose the parameters twice. I'm not sure why it's happening, but we get into DecomposeNode with a GT_LONG node. I tried simply ignoring GT_LONGs in DecomposeNode, but that lead to another failure, so I just left the NYI here. For the rest, we still need to initialize the retTypeDesc when we morph the node into a call, so that when we figure out what registers to use in genMultiRegCallStoreToLocal, it has been initialized.

**Comment by briansull** [comment]
Looks Good

File: src/jit/morph.cpp (hunk 2 above)
Comment (mikedn) [inline src/jit/morph.cpp]: >  we try to decompose the parameters twice. I'm not sure why it's happening

Hmm, does the tree happen to contain a long GT_STOREIND? I discovered when I did `CreateTemporary` that the solution that was used for handling moved embedded statements can result in double lowering/decomposition.

File: src/jit/importer.cpp [hunk 3]
```diff
@@ -7492,17 +7492,6 @@ GenTreePtr                Compiler::impFixupCallLongReturn(GenTreePtr     call,
     unsigned retRegCount = retTypeDesc->GetReturnRegCount();
     // must be a long returned in two registers
     assert(retRegCount == 2);
-
```
Comment (sivarv) [inline src/jit/importer.cpp]: Function header comment needs to be updated, since this routine is just initializing return type descriptor of a call node.  Also should we rename this method to be representative of its new functionality (e.g. impInitCallReturnTypeDesc())

Note that right now impFixupCallStructReturn() is also initializing return type desc of a call node for unix amd64.  @briansull would need to add similar code for Arm64 case.  In future when we add support for arrm32 this routine would come handy.

File: src/jit/compiler.h [hunk 4]
```diff
@@ -7992,7 +7993,10 @@ public :
         // Methods returning a struct in two registers is considered having a return value of TYP_STRUCT.
         // Such method's compRetNativeType is TYP_STRUCT without a hidden RetBufArg
         return varTypeIsStruct(info.compRetNativeType) && (info.compRetBuffArg == BAD_VAR_NUM);
-#else 
+#elif FEATURE_MULTIREG_RET && defined(_TARGET_X86_)
```
Comment (sivarv) [inline src/jit/compiler.h]: For readability, I would suggest to organize this as follows

#if FEATURE_MULTIREG_RET

#if defined(FEATURE_UNIX_AMD64_STRUCT_PASSING) || defined(_TARGET_ARM)
    return varTypeIsStruct(info.compRetNativeType) && (info.compRetBuffArg == BAD_VAR_NUM);
#else if defined (_TARGET_X86_)
    return varTypeIsLong(info.compRetNativeType);
#else 
    unreached();
#endif

#else 
   return false;
#endif

File: src/jit/flowgraph.cpp [hunk 5]
```diff
@@ -21584,7 +21584,6 @@ Compiler::fgWalkResult      Compiler::fgUpdateInlineReturnExpressionPlaceHolder(
                                        : NO_CLASS_HANDLE;
 #endif // defined(FEATURE_HFA) || defined(FEATURE_UNIX_AMD64_STRUCT_PASSING)
 
-
```
Comment (sivarv) [inline src/jit/flowgraph.cpp]: (Minor) Since the edit is just a blank line deletion, this file could be excluded from PR.

File: src/jit/lower.cpp [hunk 6]
```diff
@@ -381,8 +383,44 @@ void Lowering::DecomposeNode(GenTreePtr* pTree, Compiler::fgWalkData* data)
...
+                    {
+                        // If var wasn't promoted, we can just set lvIsMultiRegArgOrRet
+                        comp->lvaTable[varNum].lvIsMultiRegArgOrRet = true;
+                        break;
+                    }
+                }
+
+                // Otherwise, we need to force var = call()
+                GenTreePtr* treePtr = nullptr;
+                parent = tree->gtGetParent(&treePtr);
```
Comment (sivarv) [inline src/jit/lower.cpp]: It would be a good idea to add an assert that treePtr is not null after the call to gtGetParent()(

File: src/jit/lower.cpp [hunk 7]
```diff
@@ -381,8 +383,44 @@ void Lowering::DecomposeNode(GenTreePtr* pTree, Compiler::fgWalkData* data)
         }
         break;
     case GT_CALL:
-        NYI("Call with TYP_LONG return value");
-        break;
+        {
+            GenTree* parent = data->parent;
+            if (parent != nullptr)
```
Comment (sivarv) [inline src/jit/lower.cpp]: I see that we don't need to force IR to be var=call if GT_CALL is a top-level node and hence the check parent != nullptr.  It would be useful to add a comment line explaining that here.

File: src/jit/lower.cpp (hunk 1 above)
Comment (sivarv) [inline src/jit/lower.cpp]: @CarolEidt can new asgStmt become a top-level stmt since the check above is "if (parent !=nullptr)" that is GT_CALL node has a parent.  The case where GT_CALL node is a top-level stmt without any parent, there is no need to force IR to be var=call.  Did I mis-understood anything?

Comment (CarolEidt) [inline src/jit/lower.cpp]: That is a good point. So I don't think a TODO is required here. A comment would be useful to the effect that, since this is known to have a parent, we needn't worry about creating a new top-level statement.

File: src/jit/morph.cpp (hunk 2 above)
Comment (adiaaida) [inline src/jit/morph.cpp]: As far as I can tell, no, the tree does not have a GT_STOREIND. What seems to be happening is that we decompose the args to the call, then we decompose the call to be var = call(), but then we try to decompose one of the args again. I'm not sure why it's doing this, and why it only tries to do this with a node that has been morphed into a call (is there some strange ordering of events that happens that I'm not aware of?).

File: src/jit/importer.cpp (hunk 3 above)
Comment (adiaaida) [inline src/jit/importer.cpp]: updated the function name and header. Should we also change the logic of impFixupCallStructReturn to call this function (and make it available to more than just TARGET_X86)?

Comment (sivarv) [inline src/jit/importer.cpp]: Yes, I would like that change.  But I am equally fine if you don't want to club that change with this PR.

File: src/jit/lower.cpp [hunk 8]
```diff
@@ -381,8 +383,47 @@ void Lowering::DecomposeNode(GenTreePtr* pTree, Compiler::fgWalkData* data)
...
+            {
+                // We do not need to force var = call() if the call is a top-level node.
```
Comment (sivarv) [inline src/jit/lower.cpp]: Please add this comment above 388.  The fact that we are forcing IR to be var=call only if GT_CALL node has a parent indicates that GT_CALL node is not a top-level stmnt.

**Comment by sivarv** [comment]
Looks good.

**Comment by CarolEidt** [comment]
LGTM

**Comment by adiaaida** [comment]
@dotnet-bot test Windows_NT x64 Release please

**Comment by adiaaida** [comment]
@dotnet-bot test Windows_NT x64 Release

**Comment by adiaaida** [comment]
@dotnet-bot test this please

File: src/jit/morph.cpp (hunk 2 above)
Comment (mikedn) [inline src/jit/morph.cpp]: I think I ran into this issue myself. As far as I understand `TYP_LONG` call arguments aren't decomposed so a decomposed call tree can still contain `GT_LONG` nodes. This works fine until you run into a call that is itself `TYP_LONG`. Now the call tree is moved to an embedded statement and that statement will be decomposed after the current (top-level) statement is decomposed. But the tree of that embedded statement was already decomposed...

In short, we need `GTF_STMT_SKIP_DECOMPOSE` or something similar. Though I wonder why ignoring `GT_LONG` nodes in decomposition didn't work.
